modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java

```diff
@@ -2834,6 +2829,79 @@ else if (grid0.context().invalid())
             }
         }
 
+        /**
+         * @param grpCtx Cahce group.
+         * @param nodesToExclude Nodes to exclude from check.
+         * @param proposedSuppliers Map of proposed suppliers for groups.
+         * @return True if all local partition of group specified have a copy in cluster, false otherwise.
+         */
+        private boolean haveCopyLoclaPartitions(
+            CacheGroupContext grpCtx,
+            Set<UUID> nodesToExclude,
+            Map<UUID, Map<Integer, Set<Integer>>> proposedSuppliers
+        ) {
+            GridDhtPartitionFullMap fullMap = grpCtx.topology().partitionMap(false);
+
+            if (fullMap == null)
+                return false;
+
+            UUID localNodeId = grid.getLocalNodeId();
+
+            GridDhtPartitionMap localPartMap = fullMap.get(localNodeId);
+
+            int parts = grpCtx.topology().partitions();
+
+            for (int p = 0; p < parts; p++) {
+                if (localPartMap.get(p) != GridDhtPartitionState.OWNING)
+                    continue;
+
+                boolean foundCopy = false;
+
+                for (Map.Entry<UUID, GridDhtPartitionMap> entry : fullMap.entrySet()) {
+                    if (localNodeId.equals(entry.getKey()) || nodesToExclude.contains(entry.getKey()))
+                        continue;
+
+                    //Rebalance in this cache.
+                    if (entry.getValue().hasMovingPartitions())
+                        continue;
+
+                    if (entry.getValue().get(p) == GridDhtPartitionState.OWNING) {
```

Comment:
Are we safe in case of late affinity assignment issue? At first glance we ignore ideal assignment.

Reply:
I consider checking on ideal assignment only make code difficult.
We already heave a batch of test, that continue over five minutes, but no one fails due to assignment is not ideal.